### PR TITLE
[Relayminer] Query for on-chain session param `num_blocks_per_session`

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Refactors the relayminer to query for the current `num_blocks_per_session` **each (or even multiple) time(s) it's used**. This is a naive implementation to just get the relayminer using the on-chain param instead of the const that's still defined in the session keeper pkg but which will be removed in forthcoming work.

## Next Steps
2. Implement `ModuleParamsClient[P any]` (see: [notion - ModuleParamsClient](https://www.notion.so/buildwithgrove/ModuleParamsClient-4af05dff9a0b4a22b157d2d92f2b262f)).
1. Refactor `sessionQuerier` to use `ModuleParamsClient[sessiontypes.Params]` to avoid excessive querying.
1. ...

## Issue

- #517 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [x] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [x] I create and reference any new tickets, if applicable
- [x] I have left TODOs throughout the codebase, if applicable

----

| :rocket: This description was created by [Ellipsis](https://www.ellipsis.dev) for commit bf24e6c6d0e5199387598c891eacdf4c02d35e6b  | 
|--------|

### Summary:
Refactors the `relayminer` to dynamically query the `num_blocks_per_session` parameter from the blockchain, updating relevant classes to utilize this new approach.

**Key points**:
- Refactor `relayminer` to dynamically query `num_blocks_per_session`.
- Update `SessionQueryClient` interface with new methods for session parameters.
- Implement new methods in `sessionQuerier`.
- Utilize new session parameter methods in `relayerProxy` and related classes.
- Prepare for future enhancements in session handling and grace period calculations.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: testutil/testclient/testqueryclients/sessionquerier.go [hunk 1]
```diff
@@ -99,3 +116,23 @@ func AddToExistingSessions(
 		delete(sessionsMap, sessionId)
 	})
 }
+
+func GetSessionGracePeriodBlockCount(_ context.Context, sessionEndHeight int64) (int64, error) {
+	numBlocksPerSession := sessiontypes.DefaultParams().NumBlocksPerSession
+	return int64(session.GetSessionGracePeriodBlockCount(numBlocksPerSession)), nil
+}
+
+func GetSessionGracePeriodEndHeight(_ context.Context, sessionEndHeight int64) (int64, error) {
```
Comment (bryanchriswhite) [inline testutil/testclient/testqueryclients/sessionquerier.go:125]:
```suggestion
func GetSDefaultessionGracePeriodEndHeight(_ context.Context, sessionEndHeight int64) (int64, error) {
```

File: testutil/testclient/testqueryclients/sessionquerier.go [hunk 2]
```diff
@@ -99,3 +116,23 @@ func AddToExistingSessions(
 		delete(sessionsMap, sessionId)
 	})
 }
+
+func GetSessionGracePeriodBlockCount(_ context.Context, sessionEndHeight int64) (int64, error) {
```
Comment (bryanchriswhite) [inline testutil/testclient/testqueryclients/sessionquerier.go:120]:
```suggestion
func GetDefaultSessionGracePeriodBlockCount(_ context.Context, sessionEndHeight int64) (int64, error) {
```

File: testutil/testclient/testqueryclients/sessionquerier.go [hunk 3]
```diff
@@ -58,6 +59,22 @@ func NewTestSessionQueryClient(
 		}).
 		AnyTimes()
 
+	sessionQuerier.EXPECT().GetSessionGracePeriodBlockCount(gomock.Any(), gomock.Any()).
+		DoAndReturn(GetSessionGracePeriodBlockCount).
+		AnyTimes()
+
+	sessionQuerier.EXPECT().GetSessionGracePeriodEndHeight(gomock.Any(), gomock.Any()).
+		DoAndReturn(GetSessionGracePeriodEndHeight).
```
Comment (bryanchriswhite) [inline testutil/testclient/testqueryclients/sessionquerier.go:67]:
```suggestion
		DoAndReturn(GetDefaultSessionGracePeriodEndHeight).
```

File: pkg/client/interface.go [hunk 4]
```diff
@@ -274,4 +274,18 @@ type SessionQueryClient interface {
...
 	) (*sessiontypes.Session, error)
+	// GetParams queries the chain for the current session module parameters.
+	GetParams(ctx context.Context) (*sessiontypes.Params, error)
+	// GetSessionGracePeriodEndHeight returns the height at which the grace period for
+	// the session ending with sessionEndHeight elapses.
+	GetSessionGracePeriodEndHeight(ctx context.Context, sessionEndHeight int64) (int64, error)
+	// GetSessionGracePeriodBlockCount returns the number of blocks in the grace period
+	// for the session which includes queryHeight.
+	GetSessionGracePeriodBlockCount(ctx context.Context, queryHeight int64) (uint64, error)
+	// IsWithinGracePeriod returns true if the grace period for the session ending with
```
Comment (Olshansk) [inline pkg/client/interface.go:285]: Where is this being use?

File: pkg/client/interface.go [hunk 5]
```diff
@@ -274,4 +274,18 @@ type SessionQueryClient interface {
...
+	// sessionEndHeight has not yet elapsed, given currentHeight.
+	IsWithinGracePeriod(ctx context.Context, sessionEndHeight, currentHeight int64) (bool, error)
+	// IsPastGracePeriod returns true if the grace period for the session ending with
+	// sessionEndHeight has elapsed, given currentHeight.
+	IsPastGracePeriod(ctx context.Context, sessionEndHeight, currentHeight int64) (bool, error)
```
Comment (Olshansk) [inline pkg/client/interface.go:290]: Given that we have `IsWithinGracePeriod`, why do we need `IsPastGracePeriod` as well?

File: pkg/client/interface.go [hunk 6]
```diff
@@ -274,4 +274,18 @@ type SessionQueryClient interface {
...
 		blockHeight int64,
 	) (*sessiontypes.Session, error)
+	// GetParams queries the chain for the current session module parameters.
+	GetParams(ctx context.Context) (*sessiontypes.Params, error)
+	// GetSessionGracePeriodEndHeight returns the height at which the grace period for
+	// the session ending with sessionEndHeight elapses.
+	GetSessionGracePeriodEndHeight(ctx context.Context, sessionEndHeight int64) (int64, error)
+	// GetSessionGracePeriodBlockCount returns the number of blocks in the grace period
+	// for the session which includes queryHeight.
+	GetSessionGracePeriodBlockCount(ctx context.Context, queryHeight int64) (uint64, error)
```
Comment (Olshansk) [inline pkg/client/interface.go:284]: Is there a way to just have one o these?

**Review by Olshansk — CHANGES_REQUESTED** [review]
@bryanchriswhite I'm pausing the review at `pkg/client/interface.go`. 

Here is my current thinking / concern:

1. I'm not a fan of how many times `GracePeriod` is coming up in so many functions in the `SessionQueryClient`. The `Session` is a first class citizen, the Grace period is a hack for an edge case, but it's becoming a "first class citizen", which is NOT what we want.

2. Having said (1), I can get a sense of why/how we need SOME visibility to it, but I want to minimize it's surface area. For example, I see us being able to just have one of `IsWithinGracePeriod` or `IsPartGracePeriod` without having to support / maintain both endpoints.

Lmk what you think. If you believe a conversation would be faster, you know how to reach me ;)

File: pkg/client/interface.go (hunk 4 above)
Comment (bryanchriswhite) [reply pkg/client/interface.go:285]: image

**Comment by bryanchriswhite** [comment]
> 1. I'm not a fan of how many times `GracePeriod` is coming up in so many functions in the `SessionQueryClient`. The `Session` is a first class citizen, the Grace period is a hack for an edge case, but it's becoming a "first class citizen", which is NOT what we want.

Currently, `SessionGracePeriod` is in terms of sessions. It's currently being used to calculate various block durations / heights / offsets. If we rewrite it to be in terms of blocks, the need for `GetSessionGracePeriodBlockCount()` goes away.

I would propose renaming it to `SessionGracePeriodBlocks`. I would also expect this to become a param at some point.

File: pkg/client/interface.go (hunk 6 above)
Comment (bryanchriswhite) [reply pkg/client/interface.go:284]: As per https://github.com/pokt-network/poktroll/pull/538#issuecomment-2126369870, `#GetSessionGracePeriodBlockCount()` will go away (for now). Given that, I think `#GetSessionGracePeriodEndHeight()` also can continue to be a static function, instead of becoming a client method.

File: pkg/client/interface.go (hunk 5 above)
Comment (bryanchriswhite) [reply pkg/client/interface.go:290]: Probably not. Looking at the usages, it seems like it would be better to keep `IsPastGracePeriod` (maybe rename to `IsGracePeriodElapsed`) and get rid of `IsWithinGracePeriod`.

**Comment by bryanchriswhite** [comment]
@Olshansk I'm marking this as a draft for now. I think this will change now that we've decided to add the shared module. I may close and re-open this PR on a new branch or rebase this one.

File: pkg/client/interface.go [hunk 7]
```diff
@@ -275,3 +275,8 @@ type SessionQueryClient interface {
 		blockHeight int64,
 	) (*sessiontypes.Session, error)
 }
+
+type SharedQueryClient interface {
+	// GetParams queries the chain for the current session module parameters.
```
Comment (bryanchriswhite) [inline pkg/client/interface.go:280]:
```suggestion
	// GetParams queries the chain for the current shared module parameters.
```

File: pkg/client/query/sharedquerier.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+// sharedQuerier is a wrapper around the sharedtypes.QueryClient that enables the
+// querying of on-chain shared information through a single exposed method
+// which returns an sharedtypes.Session struct
+type sharedQuerier struct {
+	clientConn    grpc.ClientConn
+	sharedQuerier sharedtypes.QueryClient
+}
+
+// NewSessionQuerier returns a new instance of a client.SessionQueryClient by
```
Comment (bryanchriswhite) [inline pkg/client/query/sharedquerier.go:23]: #PUC

File: pkg/relayer/session/session.go [hunk 9]
```diff
@@ -194,7 +195,7 @@ func (rs *relayerSessionsManager) mapBlockToSessionsToClaim(
 		// downstream at the waitForEarliestCreateClaimsHeight step.
 		// TODO_BLOCKER: Introduce governance claim and proof window durations,
 		// implement off-chain window closing and on-chain window checks.
-		if !IsWithinGracePeriod(endBlockHeight, block.Height()) {
+		if !shared.IsGracePeriodElapsed(sessionEndHeight, block.Height()) {
```
Comment (bryanchriswhite) [inline pkg/relayer/session/session.go:198]:
```suggestion
		if shared.IsGracePeriodElapsed(sessionEndHeight, block.Height()) {
```

File: pkg/relayer/session/session.go [hunk 10]
```diff
@@ -194,7 +195,7 @@ func (rs *relayerSessionsManager) mapBlockToSessionsToClaim(
 		// downstream at the waitForEarliestCreateClaimsHeight step.
```
Comment (bryanchriswhite) [inline pkg/relayer/session/session.go:195]: #PUC

**Comment by red-0ne** [comment]
@Olshansk 
> 1. I'm not a fan of how many times `GracePeriod` is coming up in so many functions in the `SessionQueryClient`. The `Session` is a first class citizen, the Grace period is a hack for an edge case, but it's becoming a "first class citizen", which is NOT what we want.

I disagree on that, I see a healthy ratio of `SessionEndBlockHeight`/`GracePeriod` in this diff.

It is, one of the key times of the C&P life cycle just like `Claim/Proof` window submissions that will be almost as visible as soon as we start relying on them.

Although, I agree that the grace period is a bit special in the sense that it indicates when a session is no longer accepting new relays to it.

File: x/shared/session.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+// Returns session number 1 for block 1 to block NumBlocksPerSession - 1 (inclusive).
+// i.e. If NubBlocksPerSession == 4, session == 1 for [1, 4], session == 2 for [5, 8], etc.
+func GetSessionNumber(blockHeight int64) int64 {
+	if blockHeight <= 0 {
+		return 0
+	}
+
+	// TODO_BLOCKER(#543): If the num_blocks_per_session param has ever been changed,
+	// this function may cause unexpected behavior.
+	return ((blockHeight - 1) / NumBlocksPerSession) + 1
```
Comment (Olshansk) [inline x/shared/session.go:49]: Still using the constant here. Is there going to be another PR after this?

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just a few minor comments but the only blocking ones are in `x/shared/session.go`

File: x/shared/session.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+// NumBlocksPerSession is a place-holder that will be removed once the respective governance
+// parameter is implemented.
+//
+// TODO_BLOCKER(#517): Remove direct usage of these constants in helper functions
+// when they will be replaced by governance params
+const NumBlocksPerSession = 4
+
+// SessionGracePeriodBlocks is the number of blocks after the session ends before the
+// "session grace period" is considered to have elapsed.
+const SessionGracePeriodBlocks = 4
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/shared/session.go:12]: See: https://github.com/pokt-network/poktroll/pull/538#issuecomment-2126369870

File: x/shared/session.go (hunk 11 above)
Comment (bryanchriswhite) [reply x/shared/session.go:49]: See: https://github.com/pokt-network/poktroll/pull/538#discussion_r1616648712

File: pkg/client/query/sharedquerier.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+	querier.sharedQuerier = sharedtypes.NewQueryClient(querier.clientConn)
+
+	return querier, nil
+}
+
+// GetParams queries & returns the shared module on-chain parameters.
+//
+// TODO_TECHDEBT(#543): We don't really want to have to query the params for every method call.
+// Once `ModuleParamsClient` is implemented, use its replay observable's `#Last` method
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/client/query/sharedquerier.go]: Depends on how we prioritize it. From our discussion about historical params, I got the impression that `ModuleParamsClient` wouldn't be a high priority for the same reason. It would reduce background network overhead but until we have a reason to believe that that's significantly impacting performance, I don't see another reason to prioritize it.

**Review by red-0ne — APPROVED** [review]
Looks good to me! Many parts are waiting for #516 and #517, Eager to see how all of these will fit together.

**Review by Olshansk — APPROVED** [review]

**Comment by bryanchriswhite** [comment]
:rotating_light: Do not delete this branch until #555 changes base! :rotating_light:
